heron/scheduler-core/src/java/com/twitter/heron/scheduler/utils/SchedulerUtils.java

```diff
@@ -39,12 +40,105 @@
+
+    public enum ExecutorPortNames {
+      MASTER_PORT("master"),
+      TMASTER_CONTROLLER_PORT("tmaster-ctl"),
+      TMASTER_STATS_PORT("tmaster-stats"),
+      SHELL_PORT("shell-port"),
+      METRICS_MANAGER_PORT("metrics-mgr"),
+      SCHEDULER_PORT("scheduler"),
+      METRICS_CACHE_MASTER_PORT("metrics-cache-m"),
+      METRICS_CACHE_STATS_PORT("metrics-cache-s"),
+      CHECKPOINT_MANAGER_PORT("ckptmgr");
+
+      private final String name;
+
+      ExecutorPortNames(String name) {
+        this.name = name;
+      }
+
+      public String toString() {
+        return this.name;
+      }
+
+      public static String[] getPortNames() {
+        List<String> names = new LinkedList<>();
+        for (ExecutorPortNames executorPortNames : ExecutorPortNames.values()) {
+          names.add(executorPortNames.toString());
+        }
+
+        return (String[]) names.toArray();
+      }
+    }
+
+    public static String[] getRequiredPorts() {
+      return new String[]{
+          ExecutorPortNames.MASTER_PORT.toString(),
+          ExecutorPortNames.TMASTER_CONTROLLER_PORT.toString(),
+          ExecutorPortNames.TMASTER_STATS_PORT.toString(),
+          ExecutorPortNames.SHELL_PORT.toString(),
+          ExecutorPortNames.METRICS_MANAGER_PORT.toString(),
+          ExecutorPortNames.SCHEDULER_PORT.toString(),
+          ExecutorPortNames.METRICS_CACHE_MASTER_PORT.toString(),
+          ExecutorPortNames.METRICS_CACHE_STATS_PORT.toString(),
+          ExecutorPortNames.CHECKPOINT_MANAGER_PORT.toString()
+      };
+    }
+
+    public static int getRequiredNumOfPorts() {
+      return getRequiredPorts().length;
+    }
+
+    public static ExecutorPorts withRequiredPorts(List<Integer> portsToUse) {
+      ExecutorPorts executorPorts = new ExecutorPorts();
+      executorPorts.setRequiredPorts(portsToUse);
+      return executorPorts;
+    }
+
+    public void setRequiredPorts(List<Integer> portsToUse) {
+      if (portsToUse.size() < getRequiredNumOfPorts()) {
+        throw new IllegalArgumentException("Did not provide enough Ports. "
+            + "Number of ports provided: " + portsToUse.size()
+            + ". Number of ports required: " + getRequiredNumOfPorts());
+      }
+
+      for (int i = 0; i < portsToUse.size(); i++) {
+        int port = portsToUse.get(i);
+        if (port == -1) {
+          throw new IllegalArgumentException("Invalid Port: "
+              + port + ". Failed to find enough ports for executor");
+        } else {
+          this.put(ExecutorPortNames.values()[i], port);
+        }
+      }
+    }
+
+    public static Object getRequired(ExecutorPorts ports, ExecutorPortNames type) {
+      Object ret = getOptional(ports, type);
+      if (ret == null) {
+        throw new IllegalArgumentException("Required executor port "
+            + type.name() + " is not found");
+      }
+      return ports.get(type);
+    }
+
+    public static Object getOptional(ExecutorPorts ports, ExecutorPortNames type) {
```

Comment:
The method name and the return type do not match. Moreover could you please explain the use of this method.

Reply:
This method is just used for getting ports that are not required thus I don't throw an exception if I don't find the port in the map unlike getRequired

Reply:
I thought `Optional` referred to `java.util.Optional`.

Reply:
If we have an `ExecutorPort` enum which includes whether it's required or optional, then the caller doesn't need to know which of the two methods to call here. They can just call `getPort(ExecutorPort port)` and the method can return `Integer` or throw if a required port isn't found.